### PR TITLE
Add Mixer integration tests

### PR DESCRIPTION
**Pull request by douglas-reid**
This PR establishes a baseline set of integration tests that exercise
some core Mixer features, including:
- basic monitoring
- denials aspect
- rate limiting
- adding new telemetry via API

There is a lot of room for improvement, but this should establish a
decent baseline for verification of behavior. Other apsects, such
as whitelisting, will need to be added when support for adding
descriptors via API, etc., is supported.

Tested with `--use_local_cluster` flag on MPB with minikube.

### CODE REVIEW

File: tests/e2e/framework/istioctl.go [hunk 1]
```diff
@@ -131,6 +131,11 @@ func (i *Istioctl) CreateRule(rule string) error {
 		i.namespace, i.istioNamespace, rule))
 }
 
+// CreateMixerRule create new rule(s)
+func (i *Istioctl) CreateMixerRule(scope, subject, rule string) error {
+	return i.run(fmt.Sprintf("-n %s mixer rule create %s %s -f %s", i.namespace, scope, subject, rule))
```
Comment (ayj) [inline tests/e2e/framework/istioctl.go]: `--istioNamespace %s` is needed to specify namespace of istio system components (see CreateRule above for example).

File: tests/e2e/tests/mixer/mixer_test.go [hunk 2]
```diff
@@ -0,0 +1,579 @@
...
+	}
+
+	metricsPortFwd := fmt.Sprintf("kubectl port-forward %s %d:%d -n %s", strings.Trim(pod, "'"), mixerPrometheusPort, mixerPrometheusPort, m.namespace)
+	if m.metricsPortFwdProcess, err = util.RunBackground(metricsPortFwd); err != nil {
+		glog.Errorf("Failed to port forward: %s", err)
+		return err
+	}
+	glog.Infof("mixer metrics port-forward running background, pid = %d", m.metricsPortFwdProcess.Pid)
+
+	apiPortFwd := fmt.Sprintf("kubectl port-forward %s %d:%d -n %s", strings.Trim(pod, "'"), mixerAPIPort, mixerAPIPort, m.namespace)
```
Comment (ayj) [inline tests/e2e/tests/mixer/mixer_test.go]: Liens 123-130 shouldn't be needed anymore with https://github.com/istio/istio/pull/205.

Comment (douglas-reid) [inline tests/e2e/tests/mixer/mixer_test.go]: These should now be removed.

File: tests/e2e/framework/istioctl.go (hunk 1 above)
Comment (douglas-reid) [inline tests/e2e/framework/istioctl.go]: Done.

**Comment by douglas-reid** [comment]
Thanks for the review!  I've updated to match comments (and upgrade istioctl).

PTAL.

File: tests/e2e/tests/mixer/mixer_test.go [hunk 3]
```diff
@@ -0,0 +1,568 @@
...
+}
+
+var (
+	tc    *testConfig
+	rules = []string{rateLimitRule, denialRule, newTelemetryRule, routeAllRule, routeReviewsVersionsRule, emptyRule}
+)
+
+func (t *testConfig) Setup() error {
+	t.gateway = "http://" + tc.Kube.Ingress
+	for _, rule := range rules {
```
Comment (sebastienvas) [inline tests/e2e/tests/mixer/mixer_test.go]: If you are not modifying the rules, you can just use the original file.

File: istio.VERSION [hunk 4]
```diff
@@ -4,6 +4,6 @@ export CA_HUB="docker.io/istio"
 export CA_TAG="2017-04-13-17.43.08"
 export MIXER_HUB="docker.io/istio"
 export MIXER_TAG="2017-05-09-02.32.46"
-export ISTIOCTL_URL="https://storage.googleapis.com/istio-artifacts/dbcc933388561cdf06cbe6d6e1076b410e4433e0/artifacts/istioctl"
+export ISTIOCTL_URL="https://storage.googleapis.com/istio-artifacts/1cb69d3458079c798bd87ec54f785725246007bc/artifacts/istioctl"
```
Comment (sebastienvas) [inline istio.VERSION]: This is not a commit in master

Comment (douglas-reid) [inline istio.VERSION]: OK. We need a new ISTIOCTL_URL with the latest fix. I just found one that worked. Do you know what the latest in master is?

Comment (ayj) [inline istio.VERSION]: 621b2c4c8073dd4e570c2c77f3d6f92f7ba969b2 is the latest commit in istio.io/manager master.

File: tests/e2e/tests/mixer/mixer_test.go (hunk 3 above)
Comment (douglas-reid) [inline tests/e2e/tests/mixer/mixer_test.go]: true. this code copies to t.rulesDir, which makes it easier to reference tc.rulesDir later in the create rules blocks. i'll update the code to make that clearer.

Comment (douglas-reid) [inline tests/e2e/tests/mixer/mixer_test.go]: updated to just use os.Link() instead of copying the files around.

File: istio.VERSION (hunk 4 above)
Comment (sebastienvas) [inline istio.VERSION]: I d rather use a stable tag here. Does stable-c1b35b6 work ?

File: tests/e2e/tests/mixer/mixer_test.go [hunk 5]
```diff
@@ -0,0 +1,579 @@
...
+		src := util.GetResourcePath(filepath.Join(rulesDir, fmt.Sprintf("%s", rule)))
+		dest := filepath.Join(t.rulesDir, rule)
+		ori, err := ioutil.ReadFile(src)
+		if err != nil {
+			glog.Errorf("Failed to read original rule file %s", src)
+			return err
+		}
+		content := string(ori)
```
Comment (douglas-reid) [inline tests/e2e/tests/mixer/mixer_test.go]: section removed.

File: istio.VERSION (hunk 4 above)
Comment (douglas-reid) [inline istio.VERSION]: updated to SHA suggested.

**Comment by douglas-reid** [comment]
PTAL. Attempted to address previous comments. Changes include:
- updated SHA for istio.VERSION
- switch to os.Link for getting rules file into tc.rulesDir

Tested on minikube: passed.

**Comment by sebastienvas** [comment]
You need to fix the linters issue first tough

**Comment by sebastienvas** [comment]
You can use stable-621b2c4 tag now instead of the SHA.

**Comment by douglas-reid** [comment]
PTAL. Updated:
- fix linter warnings
- istioctl moved to stable tag

**Pull request merged**